### PR TITLE
Add support of more JIT-compiled code for profiling

### PR DESCRIPTION
**Pull request by randomstuff**
Extends [PR-1432](https://github.com/dolphin-emu/dolphin/pull/1432) in order to support [more JIT-compiled code](https://github.com/dolphin-emu/dolphin/pull/1432#issuecomment-68222071):

  * VertexLoader;
  * shared/common JIT-compiled code (`JitAsmCommon`);
  * far code;
  * dispatch loop (?) (`JitAsm`).

I don't really known what some of the JIT-compiled functions do, so the names might need to be adjusted.

For the fare code case, the code in semantically part of the basic block, so we generate a symbol name in the format `"JIT_PPC_${basicBlockAddress}/${typeOfFarCode}%${farCodeAddress}"`. This means that:

  * most profilig solutions will not be able to account them with their basic block as they will not recognize that `"JIT_PPC_${basicBlockAddress}/${typeOfFarCode}%${farCodeAddress}"` is  a part of `"JIT_PPC_${basicBlockAddress}"`;

  * we can teach FlameGraph/`stackcollapse`-based solutions to recognize this with `sed 's/%/;/m'`.

Note: the current version of `stackcollapse-perf.pl` does a `s/;/:/` in order to handle Java symbols such as `Lorg/mozilla/javascript/ContextFactory;.call(Lorg/mozilla/javascript/ContextAction;)Ljava/lang/Object;` so we can't use `"JIT_PPC_${basicBlockAddress}/${typeOfFarCode};${farCodeAddress}"` without disabling the java handling code. This is why I'm using `%`.

### CODE REVIEW

File: Source/Core/Common/StringUtil.h [hunk 1]
```diff
@@ -13,6 +13,8 @@
 
 #include "Common/Common.h"
 
+std::string StringFromFormat(const char* format, va_list args);
```
Comment (magumagu) [inline Source/Core/Common/StringUtil.h]: This is a dangerous; on some platforms, va_list is a void* or char*, so you could end up calling the wrong overload.  Please change the name.

**Comment by Tilka** [comment]
Sorry not to mention this earlier but... do you actually ever need `Register(start, size, name)`? You could just change the interface to `Register(start, end, name)` and get rid of all those casts and subtractions.

**Comment by randomstuff** [comment]
Indeed the `Register(start, end, name)` would be easier for the callee. I was initially hesistating between the two interfaces but I ended up using the `Register(start, size, name)` because

 * I thought the interface was less ambiguous (does `Register(start, end, name)` means `[a,b)` or `[a,b]`?);
 * it is more in line with APIs such as `mmap`, `madvise`;
 * because the three APIs I'm wrapping.

But it's probably better to change to `Register(start, end, name)` or add an overloaded call.

**Comment by degasus** [comment]
Indexing with pointers is always [start,end). So size is calculated by end-start. This is the only way to support a zero size with pointers.

**Comment by degasus** [comment]
Is there another way to register the far code? Maybe also per block. Just save the farcode start pointer on initialization and register the end pointer in https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/PowerPC/JitCommon/JitCache.cpp#L145

**Comment by degasus** [comment]
As PR #1887 is merged, the vertex loader jit already register itself. So this part can be skipped.

@randomstuff Through I like to know which function does use the far code, I don't like to have to write this in every instruction which uses far code. I guess if it's just "JIT_PPC_${basicBlock}_farcode", then this could be implemented at a central position (like the common block) instead of in each instruction.

**Comment by randomstuff** [comment]
OK, I'll have  a look at this.

**Comment by randomstuff** [comment]
Rebased the code. The farcode part is currently missing however.

**Comment by degasus** [comment]
@dolphin-emu-bot rebuild

LGTM

File: Source/Core/Common/JitRegister.h [hunk 2]
```diff
@@ -5,12 +5,35 @@
...
+	RegisterV(base_address, code_size, format, args);
+	va_end(args);
+}
+
+inline void Register(const void* start, const void* end,
+	const char* format, ...)
+{
+	va_list args;
+	va_start(args, format);
+	u32 code_size = (const char*) end - (const char*) start;
```
Comment (degasus) [inline Source/Core/Common/JitRegister.h]: our windows buildbot doesn't like implicit conversion from size_t to u32 ...

**Comment by degasus** [comment]
@dolphin-emu-bot rebuild

**Comment by degasus** [comment]
grr, who else hates our lint scripts? https://buildbot.dolphin-emu.org/builders/lint/builds/1000/steps/shell/logs/stdio

**Comment by randomstuff** [comment]
Hmm yes. I always forget to run the link tools locally. It'd be nice to have (in the repository) a wrapper lint script calling all the tools with the suitable arguments (doing the same thing as the lint builder).

**Comment by degasus** [comment]
@dolphin-emu-bot rebuild

**Pull request merged**